File: src/main/java/com/GoldenMine/parser/postposition/JosaStorage.java
package com.GoldenMine.parser.postposition;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class JosaStorage {
    public static final JosaStorage INSTANCE = new JosaStorage();

    private HashMap<JosaCommunity, List<IPostPosition>> josaMap = new HashMap<>();
    private List<IPostPosition> allJosaList = new ArrayList<>();
    private JosaStorage() {
        init();
    }

    private void init() {
        addJosa(new IPostPosition주어());
        addJosa(new IPostPosition목적어());
        addJosa(new IPostPosition서술어());
        addJosa(new IPostPosition접속조사());
        addJosa(new IPostPosition부사어로());
        addJosa(new IPostPosition부사어에());
        addJosa(new IPostPosition부사어에서());
    }

    public List<? extends IPostPosition> getAllJosaList() {
        return allJosaList;
    }

    public List<? extends IPostPosition> getJosaList(JosaCommunity community) {
        return josaMap.get(community);
    }

    public void addJosa(IPostPosition josa) {
        List<IPostPosition> josaList;
        if(josaMap.containsKey(josa.getJosaCommunity())) {
            josaList = josaMap.get(josa.getJosaCommunity());
        } else {
            josaList = new ArrayList<>();
            josaMap.put(josa.getJosaCommunity(), josaList);
        }

        josaList.add(josa);
        allJosaList.add(josa);
    }
}

File: src/main/java/kr/goldenmine/parser/postposition/JosaStorage.kt
package kr.goldenmine.parser.postposition

import java.util.*

object JosaStorage {

    private val josaMap = HashMap<JosaCommunity, MutableList<IPostPosition>>()
    private val allJosaList = ArrayList<IPostPosition>()

    init {
        init()
    }

    private fun init() {
        addJosa(IPostPosition서술어())
        addJosa(IPostPosition주어())
        addJosa(IPostPosition목적어())
        addJosa(IPostPosition접속조사())
        addJosa(IPostPosition부사어로())
        addJosa(IPostPosition부사어에())
        addJosa(IPostPosition부사어에서())
        addJosa(IPostPosition까지())
        addJosa(IPostPosition비교())
        addJosa(IPostPosition소유격())
        addJosa(IPostPosition분류없음())
    }

    fun getAllJosaList(): List<IPostPosition> {
        return allJosaList
    }

    fun getJosaList(community: JosaCommunity): MutableList<IPostPosition> {
        return ArrayList(josaMap[community]!!)
    }

    fun addJosa(josa: IPostPosition) {
        val josaList: MutableList<IPostPosition>
        if (josaMap.containsKey(josa.josaCommunity)) {
            josaList = josaMap[josa.josaCommunity]!!
        } else {
            josaList = ArrayList()
            josaMap[josa.josaCommunity] = josaList
        }

        josaList.add(josa)
        allJosaList.add(josa)
    }

    fun getGyukJosaListWithBojo(): List<IPostPosition> {
        val list = getJosaList(JosaCommunity.격조사)
        list.addAll(getJosaList(JosaCommunity.보조사))

        return list
    }
}
